core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateUnnest.java

```diff
@@ -0,0 +1,402 @@
+    /**
+     * This rule supports decorrelation of UnnestNode meeting certain conditions:
+     * - the UnnestNode should be based on correlation symbols, that is: either unnest correlation symbols directly,
+     * or unnest symbols produced by a projection that uses only correlation symbols.
+     * - the UnnestNode should not have any replicate symbols,
+     * - the UnnestNode should be of type INNER or LEFT,
+     * - the UnnestNode should not have a filter.
+     */
+    private static boolean isSupportedUnnest(PlanNode node, List<Symbol> correlation, Lookup lookup)
+    {
+        if (!(node instanceof UnnestNode)) {
+            return false;
+        }
+
+        UnnestNode unnestNode = (UnnestNode) node;
+        List<Symbol> unnestSymbols = unnestNode.getMappings().stream()
+                .map(UnnestNode.Mapping::getInput)
+                .collect(toImmutableList());
+        PlanNode unnestSource = lookup.resolve(unnestNode.getSource());
+        boolean basedOnCorrelation = ImmutableSet.copyOf(correlation).containsAll(unnestSymbols) ||
+                unnestSource instanceof ProjectNode && ImmutableSet.copyOf(correlation).containsAll(SymbolsExtractor.extractUnique(((ProjectNode) unnestSource).getAssignments().getExpressions()));
+
+        return isScalar(unnestNode.getSource(), lookup) &&
+                unnestNode.getReplicateSymbols().isEmpty() &&
+                basedOnCorrelation &&
+                (unnestNode.getJoinType() == INNER || unnestNode.getJoinType() == LEFT) &&
+                (unnestNode.getFilter().isEmpty() || unnestNode.getFilter().get().equals(TRUE_LITERAL));
+    }
+
+    private PlanNode rewriteNodeSequence(
+            PlanNode root,
+            List<Symbol> leftOutputs,
+            Symbol ordinalitySymbol,
+            Symbol uniqueSymbol,
+            PlanNode sequenceSource,
+            PlanNodeId correlatedUnnestId,
+            Lookup lookup,
+            PlanNodeIdAllocator idAllocator,
+            SymbolAllocator symbolAllocator)
+    {
+        // bottom of sequence reached -- attach the rewritten source node
+        if (root.getId().equals(correlatedUnnestId)) {
+            return sequenceSource;
+        }
+
+        PlanNode source = rewriteNodeSequence(lookup.resolve(getOnlyElement(root.getSources())), leftOutputs, ordinalitySymbol, uniqueSymbol, sequenceSource, correlatedUnnestId, lookup, idAllocator, symbolAllocator);
+
+        if (root instanceof EnforceSingleRowNode) {
+            // if EnforceSingleRowNode is present in the subquery, the type of rewritten UnnestNode is LEFT.
+            // to restore the semantics of EnforceSingleRowNode, it must be checked that for each input row
+            // there is at most one row in the subquery.
+            if (isScalar(source, lookup)) {
+                return source;
+            }
+            Symbol rowNumberSymbol = symbolAllocator.newSymbol("row_number", BIGINT);
+            RowNumberNode rowNumberNode = new RowNumberNode(
```

Comment:
In case we also have `Limit` we can use the `row_number` symbol introduced by the `RowNumberNode` already added. Or am I missing somethign.